src/logging.c

```diff
@@ -628,12 +667,131 @@ rcutils_ret_t rcutils_logging_format_message(
+# define COLOR_GREEN "\033[32m"
+# define COLOR_YELLOW "\033[33m"
+#endif
+
+#define SET_COLOR_WITH_SEVERITY(status, severity, color) \
+  { \
+    switch (severity) { \
+      case RCUTILS_LOG_SEVERITY_DEBUG: \
+        color = COLOR_GREEN; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_INFO: \
+        color = COLOR_NORMAL; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_WARN: \
+        color = COLOR_YELLOW; \
+        break; \
+      case RCUTILS_LOG_SEVERITY_ERROR: \
+      case RCUTILS_LOG_SEVERITY_FATAL: \
+        color = COLOR_RED; \
+        break; \
+      default: \
+        fprintf(stderr, "unknown severity level: %d\n", severity); \
+        status = RCUTILS_RET_INVALID_ARGUMENT; \
+    } \
+  }
+#ifdef WIN32
+# define IS_OUTPUT_COLORIZED(is_colorized, stream) \
+  { \
+    if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE) { \
+      is_colorized = true; \
+    } else if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE) { \
+      is_colorized = false; \
+    } else { \
+      is_colorized = _isatty(_fileno(stream)) != 0; \
+    } \
+  }
+# define SET_OUTPUT_COLOR_WITH_COLOR(status, color, handle) \
+  { \
+    if (RCUTILS_RET_OK == status) { \
+      if (!SetConsoleTextAttribute(handle, color)) { \
+        DWORD error = GetLastError(); \
+        fprintf(stderr, "SetConsoleTextAttribute failed with error code %lu.\n", error); \
+        status = RCUTILS_RET_ERROR; \
+      } \
+    } \
+  }
+# define GET_HANDLE_FROM_STREAM(status, handle, stream) \
+  { \
+    if (RCUTILS_RET_OK == status) { \
+      if (stream == stdout) { \
+        handle = GetStdHandle(STD_OUTPUT_HANDLE); \
+      } else { \
+        handle = GetStdHandle(STD_ERROR_HANDLE); \
+      } \
+      if (INVALID_HANDLE_VALUE == handle) { \
+        DWORD error = GetLastError(); \
+        fprintf(stderr, "GetStdHandle failed with error code %lu.\n", error); \
+        status = RCUTILS_RET_ERROR; \
+      } \
+    } \
+  }
+# define EXPAND(x) x
+# define APPLY(macro, ...) EXPAND(macro(__VA_ARGS__))
+# define SET_OUTPUT_COLOR_WITH_SEVERITY(status, severity, stream, output_array) \
+  { \
+    WORD color; \
+    HANDLE handle; \
+    APPLY(SET_COLOR_WITH_SEVERITY, status, severity, color) \
+    APPLY(GET_HANDLE_FROM_STREAM, status, handle, stream) \
+    APPLY(SET_OUTPUT_COLOR_WITH_COLOR, status, color, handle) \
+  }
+# define SET_STANDARD_COLOR(status, stream, output_array) \
+  { \
+    HANDLE handle; \
+    APPLY(GET_HANDLE_FROM_STREAM, status, handle, stream) \
+    APPLY(SET_OUTPUT_COLOR_WITH_COLOR, status, COLOR_NORMAL, handle) \
+  }
+#else
+# define IS_OUTPUT_COLORIZED(is_colorized, stream) \
+  { \
+    if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_ENABLE) { \
+      is_colorized = true; \
+    } else if (g_colorized_output == RCUTILS_COLORIZED_OUTPUT_FORCE_DISABLE) { \
+      is_colorized = false; \
+    } else { \
+      is_colorized = isatty(fileno(stream)) != 0; \
+    } \
+  }
```

Comment:
This macro is almost identical to the Windows implementation above. Please avoid that duplication and only customize the `isatty` call using a platform specific macro.